l1-contracts/contracts/bridge/asset-router/L2AssetRouter.sol

```diff
@@ -2,44 +2,60 @@
 
-import {AddressAliasHelper} from "../vendor/AddressAliasHelper.sol";
-import {L2ContractHelper, L2_NATIVE_TOKEN_VAULT} from "../L2ContractHelper.sol";
-import {DataEncoding} from "../common/libraries/DataEncoding.sol";
+import {IBridgehub} from "../../bridgehub/IBridgehub.sol";
+import {AddressAliasHelper} from "../../vendor/AddressAliasHelper.sol";
 
-import {EmptyAddress, InvalidCaller, AmountMustBeGreaterThanZero} from "../errors/L2ContractErrors.sol";
+import {L2_NATIVE_TOKEN_VAULT_ADDR, L2_BRIDGEHUB_ADDR, L2_BASE_TOKEN_SYSTEM_CONTRACT_ADDR} from "../../common/L2ContractAddresses.sol";
+import {L2ContractHelper} from "../../common/libraries/L2ContractHelper.sol";
+import {DataEncoding} from "../../common/libraries/DataEncoding.sol";
+import {EmptyAddress, InvalidCaller, AmountMustBeGreaterThanZero} from "../../common/L1ContractErrors.sol";
 
 /// @author Matter Labs
 /// @custom:security-contact security@matterlabs.dev
 /// @notice The "default" bridge implementation for the ERC20 tokens. Note, that it does not
 /// support any custom token logic, i.e. rebase tokens' functionality is not supported.
-contract L2AssetRouter is IL2AssetRouter, Initializable {
-    /// @dev Chain ID of Era for legacy reasons
-    uint256 public immutable ERA_CHAIN_ID;
-
-    /// @dev Chain ID of L1 for bridging reasons
-    uint256 public immutable L1_CHAIN_ID;
-
+contract L2AssetRouter is AssetRouterBase, IL2AssetRouter {
     /// @dev The address of the L2 legacy shared bridge.
     address public immutable L2_LEGACY_SHARED_BRIDGE;
 
     /// @dev The address of the L1 asset router counterpart.
     address public override l1AssetRouter;
 
     /// @dev A mapping of asset ID to asset handler address
-    mapping(bytes32 assetId => address assetHandlerAddress) public override assetHandlerAddress;
+    // mapping(bytes32 assetId => address assetHandlerAddress) public override assetHandlerAddress;
 
     /// @notice Checks that the message sender is the L1 Asset Router.
-    modifier onlyL1AssetRouter() {
-        // Only the L1 Asset Router counterpart can initiate and finalize the deposit.
-        if (AddressAliasHelper.undoL1ToL2Alias(msg.sender) != l1AssetRouter) {
-            revert InvalidCaller(msg.sender);
+    modifier onlyAssetRouterCounterpart(uint256 _originChainId) {
+        if (_originChainId == L1_CHAIN_ID) {
+            // Only the L1 Asset Router counterpart can initiate and finalize the deposit.
+            if (AddressAliasHelper.undoL1ToL2Alias(msg.sender) != l1AssetRouter) {
+                revert InvalidCaller(msg.sender);
+            }
+        } else {
+            revert InvalidCaller(msg.sender); // xL2 messaging not supported for now
+        }
+        _;
+    }
+
+    /// @notice Checks that the message sender is the L1 Asset Router.
+    modifier onlyAssetRouterCounterpartOrSelf(uint256 _originChainId) {
+        if (msg.sender == address(this)) {} else if (_originChainId == L1_CHAIN_ID) {
```

Comment:
the only place where this modifier is used is `function finalizeDeposit`. I think the part where `self` calls it could be replaced with just an internal call. This is cheaper + cleaner

Reply:
Its because of calldata vs memory distinction. If we make it internal everything will have to be memory.

Reply:
no, internal methods can accept calldata params

Reply:
okay, then please remove the `if (msg.sender == address(this)) {}` part, the error when not self is called is handled below